src/com/amazon/corretto/crypto/provider/EvpKey.java

```diff
@@ -0,0 +1,195 @@
+        byte[] encodedParams = use(EvpKey::getDerEncodedParams);
+        try {
+            AlgorithmParameters params = AlgorithmParameters.getInstance(type.jceName);
+            params.init(encodedParams);
+            return params.getParameterSpec(paramSpec);
+        } catch (final GeneralSecurityException | IOException ex) {
+            throw new RuntimeCryptoException("Unable to deserialize parameters: "
+                + Base64.getEncoder().encodeToString(encoded), ex);
+        }
+    }
+
+    /**
+     * This method will be called by @{link #destroy()} after calling @{code internalKey.release()}.
+     */
+    protected synchronized void destroyJavaState() {
+        // NOP
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        // We try to avoid comparing the encoded values
+        // because it may be slow and may pull secret data into the Java heap
+        if (this == obj) {
+            return true;
+        }
+        if (!(obj instanceof Key)) { // Implicit null check
+            return false;
+        }
+        final Key other = (Key) obj;
+        if (!getAlgorithm().equalsIgnoreCase(other.getAlgorithm())) {
+            return false;
+        }
+
+        final byte[] otherEncoded;
+        if (obj.getClass().equals(getClass())) {
+            // If it is also an EvpKey then we can see if the internal key is the same
+            EvpKey evpOther = (EvpKey) obj;
+            if (internalKey.equals(evpOther.internalKey)) {
+                return true;
+            }
+            // If it is also an EvpKey then we can grab the other encoded value without copying it
+            evpOther.initEncoded();
+            otherEncoded = evpOther.encoded;
+        } else {
+            otherEncoded = other.getEncoded();
+        }
+
+        // Constant time equality check
+        initEncoded();
+        return MessageDigest.isEqual(encoded, otherEncoded);
+    }
+
+    @Override
+    public int hashCode() {
+        // TODO: Consider ways to avoid exposing the entire encoded object ot Java for private keys just for a hashCode
+        if (cachedHashCode == null) {
+            synchronized (this) {
+                initEncoded();
+                // Selected to match implementations of sun.security.pkcs.PKCS8Key and sun.security.x509.X509Key
+                int workingValue = 0;
+                if (isPublicKey) {
+                    workingValue = encoded.length;
+                    for (final byte b : encoded) {
+                        workingValue += (b & 0xff) * 37;
+                    }
+                } else {
+                    if (Utils.getJavaVersion() >= 17) {
+                        workingValue = Arrays.hashCode(encoded);
+                    } else {
+                        for (int idx = 0; idx < encoded.length; idx++) {
+                            workingValue += encoded[idx] * idx;
+                        }
```

Comment:
ok, that makes sense. can we trust that the JDK's `hashCode` implementations will be stable moving forward? unless there's a strong guarantee of stability, i'd suggest either one of:

1. somehow casting + calling the JDK's hashCode implementation directly here
2. adding a unit test that asserts hashCode equivalence between JDK/this impl. over a few example keys

Reply:
Hopefully it will be.

That said, BouncyCastle doesn't even try to match ([RSA Public](https://github.com/bcgit/bc-java/blob/master/prov/src/main/java/org/bouncycastle/jce/provider/JCERSAPublicKey.java#L96-L99), [EC Private](https://github.com/bcgit/bc-java/blob/master/prov/src/main/java/org/bouncycastle/jce/provider/JCEECPrivateKey.java#L394-L397)), so maybe we shouldn't bother trying either?

1. Not possible
2. We already have this.

Reply:
> BouncyCastle doesn't even try to match ([RSA Public](https://github.com/bcgit/bc-java/blob/master/prov/src/main/java/org/bouncycastle/jce/provider/JCERSAPublicKey.java#L96-L99), [EC Private](https://github.com/bcgit/bc-java/blob/master/prov/src/main/java/org/bouncycastle/jce/provider/JCEECPrivateKey.java#L394-L397)), so maybe we shouldn't bother trying either?

eh, seems like matching JDK's implementation would be path of least surprise. maybe some applications implicitly depend on that behavior, and us diverging from it would unnecessarily induce an incompatibility w/ ACCP.

> We already have this.

perfect!